Run the GC on function return and remove GC Header from Objects

lpereira opened the pull request:
I was reading your blog post about Nederlang, specifically the one where you convert it from C and optimize it, and while playing with over the weekend it I found a way to make the GC a bit more efficient.

By using a `bitvec::BitVec` to mark reachable objects instead of the `Header` in every `Object`, I was able to make things quite a bit more efficient.  Each commit have details.

lpereira committed:
Remove GC Header from Objects

By using a bitmap to store which indexes were marked in our GC heap, we
can remove the Header from each object.  As a result, both the mark and
sweep steps can be performed more efficiently:

- Unreachable ranges are drained in bulk.  Draining won't change a Vec
  capacity, so this won't trigger reallocations and potential copies.

- No need to read or write the header of every object to determine if
  it was marked.  This improves data locality and reduces memory
  traffic.

- The objects Vec may be shrink to fit if the capacity goes beyond a
  certain threshold, but keep the overallocation under a reasonable
  limit.

lpereira committed:
Run the GC on function return

Even with GC enabled on function returns, which is exercised
extensively with the fib(24) benchmark, this makes the runtime only
around 1% slower.  Compare with the unmodified GC, also being
executed on function returns, that made fib(25) 25% slower.

The parser benchmark is around 20% slower, though, so I got to look
into it.

dannyvankooten commented:
That looks awesome @lpereira - I'm hoping to be able to take a closer look some time next week.

lpereira committed:
Root objects don't need to be passed as mut borrows

Since objects themselves aren't modified anymore -- save from them
being freed if not marked during a collection cycle -- there's no need
to pass mutable borrows to the garbage collector anymore.

lpereira committed:
Change size of the bitmap as we trace/untrace objects

Since collection happens more often than tracing/untracing objects,
make it so that we resize the bitmap as we go during tracing/untracing
operations rather than during a full mark/sweep cycle.

lpereira committed:
Don't recurse into Arrays if it's present in multiple roots

Also, use the *_unchecked() variants for marking a bit in the bitmap,
as it only panics if the index we're marking goes beyond the usable
size in the allocation, but this can't happen as we know the size of
the bitmap prior to this whole operation.

lpereira committed:
Simplify the sweep step by iterating over zeros in the bitmap

Instead of trying to compute ranges to drain, which complicates things
quite a bit, iterate over the unmarked objects to free them. This makes
the sweep step incredibly simple -- and won't need as many comments
trying to explain how it works.

lpereira commented:
No rush! I ended up pushing some new changes after I sent the PR, further simplifying some things (i.e. the collection step is a lot nicer now!). I don't think I'll be making any other changes in this PR, though.

dannyvankooten commented:
So I finally had some time; although my Rust was once again quite rusty (ba dum tss) after a year of solely working in C and C++... But, this is very clever! Thanks for sharing that improvement here. :pray:
